File: src/lib/candidate-ranker.ts
"use client";

/**
 * 候选打分排序 —— 把多路召回的 Candidate[] 排出 Top N，喂给 Playlist Planner LLM。
 *
 * 不让 LLM 自己排：
 *   1) LLM 看几百首歌的元数据成本高，且会随机改顺序
 *   2) 本地公式可解释、可调参，能跟着行为日志微调
 *
 * 公式：
 *   score = weighted relevance
 *         - recent play/recommendation penalties
 *
 * 各分量来自召回阶段已经算好的 sourceScores，避免重复计算。
 * avoid 命中直接清零（不是降权，是排除）。
 */

import type { Candidate } from "./candidate-recall";
import type { MusicIntent } from "./music-intent";
import type { RecentPlayContext } from "./behavior-log";
import type { RecentRecommendationContext } from "./recommendation-log";
import { normalizeArtistKeyForLog } from "./recommendation-log";
import type { TrackInfo } from "./tauri";
import {
  dedupeSimilarTracks,
  queryAsksForSpecificVersion,
  queryExplicitlyMentionsTrack,
} from "./track-dedupe";
import { passesHardConstraints } from "./tag-recall";

export type RankedCandidate = Candidate & {
  finalScore: number;
  /**
   * 分位 bucket(0..3,0 最高分位)。
   * 续杯端按 bucket 跨段抽样,保证"主调 + 冷门"始终混合,而不是把最热的都抽完
   * 再去碰边缘候选。
   */
  bucket: 0 | 1 | 2 | 3;
  /** 各分量分解 —— 调试 / 调试日志用 */
  breakdown: {
    intent: number;
    tag: number;
    semantic: number;
    taste: number;
    behavior: number;
    transition: number;
    freshness: number;
    recentPlayPenalty: number;
    recentRecommendationPenalty: number;
    artistFatiguePenalty: number;
  };
};

export type RankerOptions = {
  /** 取 Top N 喂给 LLM。默认 100 */
  topN?: number;
  recentPlay?: RecentPlayContext;
  recentRecommendation?: RecentRecommendationContext;
};

const DEFAULT_TOP_N = 100;

// 权重思路（v2,2026-05 重构):
//   - INTENT(用户当前句子直接点的)=0.30 不变,点名了就该听到
//   - TASTE(画像派生)从 0.25 提到 0.40,**这一档现在不止 artist** ——
//     内部走 max(profile_tags, acoustic, profile/artist*0.5, audio*0.7)。
//     profile_tags(风格/情绪/年代/文化) 和 acoustic(BPM/响度/音色)替代
//     原本"全靠 topArtists"的 taste 路径,所以 Taylor 粉也能听到 Phoebe
//     Bridgers / Lorde / Mitski 这类口味邻居。
//   - TRANSITION / FRESHNESS / BEHAVIOR 保持小权重
const W_INTENT = 0.30;
const W_TASTE = 0.40;
const W_BEHAVIOR = 0.10;
const W_TRANSITION = 0.13;
const W_FRESHNESS = 0.07;

export function rankCandidates(
  candidates: Candidate[],
  intent: MusicIntent,
  options: RankerOptions = {},
): RankedCandidate[] {
  const avoidKeys = intent.musicHints.avoid
    .map((s) => s.toLowerCase().trim())
    .filter(Boolean);

  const ranked: RankedCandidate[] = [];
  for (const c of candidates) {
    // avoid 短路：歌名 / 艺人 / 专辑里出现 avoid 关键字 → 直接剔除
    if (avoidKeys.length > 0 && hitsAvoid(c, avoidKeys)) continue;
    const explicitlyMentioned = queryExplicitlyMentionsTrack(c.track, intent);
    if (
      c.semanticProfile &&
      !explicitlyMentioned &&
      !passesHardConstraints(c.semanticProfile, intent.hardConstraints)
    ) {
      continue;
    }
    if (
      options.recentPlay?.last24hTrackIds.has(c.track.id) &&
      !explicitlyMentioned
    ) {
      continue;
    }
    if (
      options.recentRecommendation?.last24hTrackIds.has(c.track.id) &&
      !explicitlyMentioned
    ) {
      continue;
    }

    const intentScore = clamp01(c.sourceScores.text ?? 0);
    const tagScore = clamp01((c.sourceScores.tag ?? 0) / 1.6);
    const semanticScore = clamp01(
      Math.max(c.sourceScores.semantic ?? 0, (c.sourceScores.semantic_broad ?? 0) * 0.65),
    );
    // taste 内部:profile_tags(风格维度)和 acoustic(声学维度)是主信号,
    // profile/artist 降到 0.5×,audio 0.7×。这就是"贴画像不贴艺人"的落点。
    const tasteScore = clamp01(
      Math.max(
        (c.sourceScores.profile_tags ?? 0) / 1.2,
        (c.sourceScores.acoustic ?? 0) * 0.9,
        (c.sourceScores.profile ?? 0) * 0.5,
        (c.sourceScores.audio ?? 0) * 0.7,
      ),
    );
    const behaviorScore = clamp01((c.sourceScores.behavior ?? 0) / 1.5);
    const transitionScore = clamp01(c.sourceScores.transition ?? 0);
    // freshness：explore 路命中 = 高新鲜度；text/behavior 命中 = 低新鲜度
    const freshnessScore = clamp01(
      (c.sourceScores.explore ?? 0) -
        (c.sourceScores.behavior ?? 0) * 0.3,
    );

    const weights = rankWeights(intent);
    // 随机扰动 0.025 → 0.08:同 query 重复触发时,让"几乎同分"的歌互相错位。
    // 配合下方的分位 bucket 洗牌,共同破"顺序固定"的体验。
    const baseScore =
      intentScore * W_INTENT +
      tagScore * weights.tag +
      semanticScore * weights.semantic +
      tasteScore * weights.taste +
      behaviorScore * weights.behavior +
      transitionScore * W_TRANSITION +
      freshnessScore * W_FRESHNESS +
      Math.random() * 0.08;

    const recentPlayPenalty = explicitlyMentioned
      ? 0
      : recentPenalty(c.track.id, options.recentPlay, 0.5, 0.25);
    const recentRecommendationPenalty = explicitlyMentioned
      ? 0
      : recentPenalty(c.track.id, options.recentRecommendation, 0.35, 0.15);
    // artist-level fatigue:近 24h / 7d 这个艺人被推过几次,推太多就降权。
    // 这是修"自由推荐总是同一个歌手"的核心 —— 即使艺人在 topArtists 里得高分,
    // 连续推 5-10 次后会被强制压下去,让池子换换人。
    // explicitlyMentioned 跳过(用户主动点的就别 fatigue 自己点的人)。
    const artistFatiguePenalty = explicitlyMentioned
      ? 0
      : computeArtistFatigue(c.track, options.recentRecommendation);
    const finalScore =
      baseScore -
      recentPlayPenalty -
      recentRecommendationPenalty -
      artistFatiguePenalty;

    console.debug("[recommend-rank]", {
      track: c.track.name,
      baseScore,
      recentPlayPenalty,
      recentRecommendationPenalty,
      finalScore,
      sourceScores: c.sourceScores,
      semantic: c.semanticProfile
        ? {
            language: c.semanticProfile.language.primary,
            region: c.semanticProfile.region.primary,
            genres: c.semanticProfile.style.genres,
            moods: c.semanticProfile.vibe.moods,
            scenes: c.semanticProfile.vibe.scenes,
          }
        : null,
    });

    if (finalScore <= 0) continue;

    ranked.push({
      ...c,
      finalScore,
      bucket: 0, // 占位,排序后按分位重写
      breakdown: {
        intent: intentScore,
        tag: tagScore,
        semantic: semanticScore,
        taste: tasteScore,
        behavior: behaviorScore,
        transition: transitionScore,
        freshness: freshnessScore,
        recentPlayPenalty,
        recentRecommendationPenalty,
        artistFatiguePenalty,
      },
    });
  }

  ranked.sort((a, b) => b.finalScore - a.finalScore);
  const deduped = queryAsksForSpecificVersion(intent)
    ? ranked
    : dedupeSimilarTracks(ranked);

  // 分位 bucket 洗牌:把 ranked 切成 4 段(top25 / 25-50 / 50-75 / 75-100),
  // 每段内 Fisher-Yates 洗牌,段间保留优先级。这样同 query 重复触发时,顺序
  // 在每段内都会变,但"最贴口味"的歌仍然排在前面。
  // queryAsksForSpecificVersion 时不洗 —— 用户点名了具体版本,排序应稳定。
  if (!queryAsksForSpecificVersion(intent) && deduped.length > 8) {
    shuffleByBuckets(deduped);
  } else {
    // 至少给每首歌打上 bucket 标签,续杯端要用
    for (let i = 0; i < deduped.length; i++) {
      deduped[i].bucket = pickBucket(i, deduped.length);
    }
  }

  return deduped.slice(0, options.topN ?? DEFAULT_TOP_N);
}

function pickBucket(rank: number, total: number): 0 | 1 | 2 | 3 {
  const ratio = rank / Math.max(1, total);
  if (ratio < 0.25) return 0;
  if (ratio < 0.5) return 1;
  if (ratio < 0.75) return 2;
  return 3;
}

function shuffleByBuckets(items: RankedCandidate[]): void {
  // 先标记每首歌的 bucket(按原排序)
  for (let i = 0; i < items.length; i++) {
    items[i].bucket = pickBucket(i, items.length);
  }
  // 按 bucket 切片,段内洗牌,再拼回去(原地修改)
  const groups: RankedCandidate[][] = [[], [], [], []];
  for (const it of items) {
    groups[it.bucket].push(it);
  }
  for (const g of groups) fisherYates(g);
  let idx = 0;
  for (const g of groups) {
    for (const it of g) items[idx++] = it;
  }
}

function fisherYates<T>(arr: T[]): void {
  for (let i = arr.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [arr[i], arr[j]] = [arr[j], arr[i]];
  }
}

function clamp01(x: number): number {
  return x < 0 ? 0 : x > 1 ? 1 : x;
}

function hitsAvoid(c: Candidate, avoidKeys: string[]): boolean {
  const haystack = [
    c.track.name.toLowerCase(),
    c.track.artists.map((a) => a.name.toLowerCase()).join(" "),
    c.track.album?.name.toLowerCase() ?? "",
  ].join(" ");
  return avoidKeys.some((k) => haystack.includes(k));
}

function recentPenalty(
  trackId: number,
  context: RecentPlayContext | RecentRecommendationContext | undefined,
  last24hPenalty: number,
  last7dPenalty: number,
): number {
  if (!context) return 0;
  if (context.last24hTrackIds.has(trackId)) return last24hPenalty;
  if (context.last7dTrackIds.has(trackId)) return last7dPenalty;
  return 0;
}

/**
 * artist-level fatigue:近期某艺人被推过太多次时降权。
 *
 * 阈值 v2(2026-05 放宽):
 *   - 24h ≥ 8 次  → -0.15 (轻扣,提示该换换人但不强制)
 *   - 24h ≥ 15 次 → 再 -0.20,累计 -0.35
 *   - 7d  ≥ 30 次 → 再 -0.15,累计 -0.50
 *
 * 之前 24h ≥5 -0.30 太狠 —— 用户喜欢 Taylor 的话,听 5 首就把后续 Taylor 压
 * 下去会导致"浮上来的歌偏冷门"。放宽到 ≥8 起步、扣得也更轻,让"用户爱听
 * 的艺人继续推"空间更大,fatigue 仍能挡住"连推 20 首同人"的极端情况。
 */
function computeArtistFatigue(
  track: TrackInfo,
  ctx: RecentRecommendationContext | undefined,
): number {
  if (!ctx) return 0;
  const key = normalizeArtistKeyForLog(track.artists[0]?.name ?? "");
  if (!key) return 0;
  const c24 = ctx.last24hArtistCounts.get(key) ?? 0;
  const c7 = ctx.last7dArtistCounts.get(key) ?? 0;
  let penalty = 0;
  if (c24 >= 8) penalty += 0.15;
  if (c24 >= 15) penalty += 0.20;
  if (c7 >= 30) penalty += 0.15;
  return penalty;
}

function rankWeights(intent: MusicIntent): {
  tag: number;
  semantic: number;
  taste: number;
  behavior: number;
} {
  const hardCount =
    intent.hardConstraints.languages.length +
    intent.hardConstraints.regions.length +
    intent.hardConstraints.genres.length +
    intent.hardConstraints.subGenres.length +
    intent.hardConstraints.vocalTypes.length +
    intent.hardConstraints.excludeLanguages.length +
    intent.hardConstraints.excludeGenres.length +
    intent.hardConstraints.excludeTags.length;
  const vibeCount =
    intent.softPreferences.moods.length +
    intent.softPreferences.scenes.length +
    intent.softPreferences.textures.length +
    (intent.softPreferences.energy === "any" ? 0 : 1);

  if (hardCount > 0) {
    // 用户明确点了硬标签:tag 路主导,taste 给一点(让画像维度也参与)
    return { tag: 0.34, semantic: 0.22, taste: 0.18, behavior: 0.06 };
  }
  if (vibeCount > 0) {
    // 用户给了氛围词:semantic + taste 平分(taste 现在内部就是画像 tags + acoustic)
    return { tag: 0.16, semantic: 0.30, taste: 0.30, behavior: 0.08 };
  }
  // 自由推荐:taste 当家(它内部是 profile_tags/acoustic/artist/audio)
  return { tag: 0.08, semantic: 0.14, taste: W_TASTE, behavior: W_BEHAVIOR };
}

File: android-native/app/src/main/java/app/pipo/nativeapp/data/CandidateRanker.kt
package app.pipo.nativeapp.data

/**
 * 候选打分排序 —— 镜像 src/lib/candidate-ranker.ts。
 *
 * 把多路 Candidate[] 按 weighted relevance + recent penalties 排出 Top N，
 * 喂给后续的 SmoothQueue 或者 Pet 的 initialBatch。
 */
object CandidateRanker {

    data class Ranked(
        val candidate: CandidateRecall.Candidate,
        val finalScore: Double,
        /** 分位 bucket(0..3,0 最高分位),供续杯端跨段抽样 */
        val bucket: Int = 0,
    )

    data class Options(
        val topN: Int = 100,
        val recentPlay: BehaviorLog.RecentPlay? = null,
        val recentRecommendation: RecommendationLog.RecentContext? = null,
        val behaviorPreference: BehaviorPreferenceSnapshot = BehaviorPreferenceSnapshot.Empty,
    )

    // 权重 v2(2026-05) —— 跟 Web 端 candidate-ranker.ts 对齐。
    // TASTE 0.25 → 0.40,内部走 max(profileTags/1.2, acoustic*0.9, profile*0.25, audio*0.7),
    // 让画像维度(风格/情绪/年代/文化/声学)替代旧版"只看 topArtists"的主信号。
    private const val W_INTENT = 0.30
    // 接歌适配只做同档候选的 tie-breaker，不能压过用户需求/口味画像。
    private const val W_TRANSITION = 0.12
    private const val W_FRESHNESS = 0.07
    private const val W_TASTE = 0.40
    private const val W_BEHAVIOR = 0.10

    fun rank(
        candidates: List<CandidateRecall.Candidate>,
        intent: PetIntent,
        options: Options = Options(),
    ): List<Ranked> {
        val avoidKeys = (intent.avoidWords + intent.aiAvoidStyles)
            .map { it.lowercase().trim() }
            .filter { it.isNotBlank() }
        val excludedArtistKeys = intent.excludeArtists
            .map(::normalizeArtistKey)
            .filter { it.isNotBlank() }
        val ranked = ArrayList<Ranked>()
        val rnd = java.util.Random()

        for (c in candidates) {
            if (avoidKeys.isNotEmpty() && hitsAvoid(c, avoidKeys)) continue
            if (excludedArtistKeys.isNotEmpty() && hitsExcludedArtist(c.track, excludedArtistKeys)) continue
            val explicitlyMentioned = TrackDedupe.queryExplicitlyMentions(
                c.track,
                intent.hardArtists, intent.hardTracks,
                intent.textArtists, intent.textTracks,
            )
            if (c.semanticProfile != null && !explicitlyMentioned &&
                !TagRecall.passesHardConstraints(c.semanticProfile, intent)
            ) continue
            val neteaseId = c.track.neteaseId
            if (neteaseId != null && options.recentPlay?.last24hTrackIds?.contains(neteaseId) == true && !explicitlyMentioned) continue

            val intentScore = clamp01(c.sourceScores[CandidateRecall.Source.Text] ?: 0.0)
            val tagScore = clamp01((c.sourceScores[CandidateRecall.Source.Tag] ?: 0.0) / 1.6)
            val semanticScore = clamp01(maxOf(
                c.sourceScores[CandidateRecall.Source.Semantic] ?: 0.0,
                (c.sourceScores[CandidateRecall.Source.SemanticBroad] ?: 0.0) * 0.65,
            ))
            // taste 内部:ProfileTags(风格维度) + Acoustic(声学维度)主导,Profile(artist) 0.25×
            val tasteScore = clamp01(maxOf(
                maxOf(
                    (c.sourceScores[CandidateRecall.Source.ProfileTags] ?: 0.0) / 1.2,
                    (c.sourceScores[CandidateRecall.Source.Acoustic] ?: 0.0) * 0.9,
                ),
                maxOf(
                    (c.sourceScores[CandidateRecall.Source.Profile] ?: 0.0) * 0.25,
                    (c.sourceScores[CandidateRecall.Source.Audio] ?: 0.0) * 0.7,
                ),
            ))
            val behaviorScore = clamp01((c.sourceScores[CandidateRecall.Source.Behavior] ?: 0.0) / 1.5)
            val sourcePreferenceDelta = (c.sourceScores[CandidateRecall.Source.PreferenceDelta] ?: 0.0) / 1.4
            val scoredPreferenceDelta = options.behaviorPreference.scoreTrack(c.track, c.semanticProfile, c.features)
            val rawPreferenceDelta = if (sourcePreferenceDelta > 0.0) {
                maxOf(sourcePreferenceDelta, scoredPreferenceDelta)
            } else {
                scoredPreferenceDelta
            }
            val preferenceDeltaScore = rawPreferenceDelta.coerceIn(-1.0, 1.0)
            val behaviorAffinityScore = clamp01(maxOf(behaviorScore, preferenceDeltaScore))
            val transitionScore = clamp01(c.sourceScores[CandidateRecall.Source.Transition] ?: 0.0)
            val freshnessScore = clamp01(
                (c.sourceScores[CandidateRecall.Source.Explore] ?: 0.0) -
                    (c.sourceScores[CandidateRecall.Source.Behavior] ?: 0.0) * 0.3
            )

            val w = rankWeights(intent)
            // 扰动 0.025 → 0.08(跟 Web 对齐),配合 bucket 洗牌共破"同 query 总同结果"
            val baseScore = intentScore * W_INTENT +
                tagScore * w.tag + semanticScore * w.semantic +
                tasteScore * w.taste + behaviorAffinityScore * w.behavior +
                transitionScore * W_TRANSITION + freshnessScore * W_FRESHNESS +
                rnd.nextDouble() * 0.08

            val recentPlayPenalty = if (explicitlyMentioned) 0.0
                else recentPenalty(neteaseId, options.recentPlay?.last24hTrackIds, options.recentPlay?.last7dTrackIds, 0.5, 0.25)
            val recentRecPenalty = if (explicitlyMentioned) 0.0
                else recentPenalty(neteaseId, options.recentRecommendation?.last24hTrackIds, options.recentRecommendation?.last7dTrackIds, 0.55, 0.24)
            // artist-level fatigue —— 修"自由推荐总是同一歌手"。
            // 阈值: 24h ≥5 次 -0.30, ≥10 -0.30(累计), 7d ≥20 -0.20。explicit 时不 fatigue。
            val artistFatiguePenalty = if (explicitlyMentioned) 0.0
                else computeArtistFatigue(c.track, options.recentRecommendation)

            val preferenceDeltaAdjust = if (explicitlyMentioned) {
                maxOf(0.0, preferenceDeltaScore) * 0.06
            } else {
                preferenceDeltaScore * 0.16
            }
            // 能量方向罚分 —— 用户明确要安静/低能量(或嗨/高能量)时，声学能量严重不符的歌降权。
            // 修"来点安静的却端出最吵的最爱"：taste 信号再强也不该违背当下能量诉求。点名要的不罚。
            val energyMismatch = if (explicitlyMentioned) 0.0
                else energyDirectionPenalty(intent, c.features)
            val finalScore = baseScore + preferenceDeltaAdjust -
                recentPlayPenalty - recentRecPenalty - artistFatiguePenalty - energyMismatch
            if (finalScore <= 0) continue
            ranked.add(Ranked(c, finalScore))
        }

        ranked.sortByDescending { it.finalScore }
        val asksVersion = TrackDedupe.queryAsksForSpecificVersion(
            intent.musicHintsGenres + intent.musicHintsMoods + intent.textTracks + intent.textArtists + intent.textAlbums
        )
        val deduped = if (asksVersion) ranked else dedupeByCandidate(ranked)

        // 分位 bucket 洗牌(同 query 重复触发时让顺序在每段内变化,跨段保持优先级)
        val withBuckets = if (!asksVersion && deduped.size > 8) {
            shuffleByBuckets(deduped)
        } else {
            // 至少打 bucket 标签
            deduped.mapIndexed { idx, r -> r.copy(bucket = pickBucket(idx, deduped.size)) }
        }
        return withBuckets.take(options.topN)
    }

    private fun pickBucket(rank: Int, total: Int): Int {
        val ratio = rank.toDouble() / maxOf(1, total)
        return when {
            ratio < 0.25 -> 0
            ratio < 0.50 -> 1
            ratio < 0.75 -> 2
            else -> 3
        }
    }

    private fun shuffleByBuckets(items: List<Ranked>): List<Ranked> {
        val withBucket = items.mapIndexed { idx, r -> r.copy(bucket = pickBucket(idx, items.size)) }
        val groups = Array(4) { ArrayList<Ranked>() }
        for (r in withBucket) groups[r.bucket].add(r)
        for (g in groups) g.shuffle()
        val out = ArrayList<Ranked>(items.size)
        for (g in groups) out.addAll(g)
        return out
    }

    /**
     * artist-level fatigue —— 镜像 Web 端 computeArtistFatigue。
     *
     * 阈值 v2(2026-05 放宽):用户反馈"压完同艺人浮上来的歌偏冷门",原来 24h ≥5 -0.30
     * 触发太早。放宽到 24h ≥8 -0.15 / ≥15 -0.20(累计 -0.35) / 7d ≥30 -0.15(累计 -0.50)。
     * 让"用户喜欢的艺人继续推"的空间更大,fatigue 仍能挡住"连推 20 首同人"的极端情况,
     * 但不会把"听了 5 首 Taylor"就立刻切到非 Taylor 的歌。
     */
    private fun computeArtistFatigue(
        track: NativeTrack,
        ctx: RecommendationLog.RecentContext?,
    ): Double {
        if (ctx == null) return 0.0
        val key = normalizeArtistKey(track.artist.split('/', '&', ',', '、').firstOrNull().orEmpty())
        if (key.isEmpty()) return 0.0
        val c24 = ctx.last24hArtistCounts[key] ?: 0
        val c7 = ctx.last7dArtistCounts[key] ?: 0
        var penalty = 0.0
        if (c24 >= 8) penalty += 0.15
        if (c24 >= 15) penalty += 0.20
        if (c7 >= 30) penalty += 0.15
        return penalty
    }

    /**
     * 能量方向罚分 —— 只在用户明确给了能量方向(softEnergy / musicHintsEnergy != any)时生效。
     * 用声学能量 eMid=(introEnergy+outroEnergy)/2 衡量：要安静(low)时越吵罚越多，要嗨(high)时
     * 越蔫罚越多。没有声学特征的歌中性(0 罚，不奖不惩)，避免误伤未分析曲目。
     * 上限 0.35：足以把"最常听的吵歌"挤下安静请求的队首，又不至于一刀切。
     */
    private fun energyDirectionPenalty(intent: PetIntent, features: AudioFeatures?): Double {
        if (features == null) return 0.0
        val target = intent.musicHintsEnergy.takeIf { it != "any" }
            ?: intent.softEnergy.takeIf { it != "any" }
            ?: return 0.0
        val eMid = (features.introEnergy + features.outroEnergy) / 2.0
        return when (target) {
            "low", "mid_low" -> (((eMid - 0.32) / 0.40).coerceIn(0.0, 1.0)) * 0.35
            "high", "mid_high" -> (((0.42 - eMid) / 0.40).coerceIn(0.0, 1.0)) * 0.30
            else -> 0.0
        }
    }

    private fun normalizeArtistKey(s: String): String =
        s.lowercase().replace(Regex("[\\s'\"`·・\\-－—_,，。.、!?！？]+"), "")

    private fun hitsExcludedArtist(track: NativeTrack, excludedArtistKeys: List<String>): Boolean {
        val whole = normalizeArtistKey(track.artist)
        val parts = track.artist
            .split("/", "&", ",", "、", " feat.", " feat ", "feat.", "feat ", " featuring ", " ft.", " ft ")
            .map(::normalizeArtistKey)
            .filter { it.isNotBlank() }
        return excludedArtistKeys.any { key ->
            parts.any { part -> part == key || (part.length >= 3 && key.length >= 3 && (part in key || key in part)) } ||
                (whole.length >= 3 && key.length >= 3 && key in whole)
        }
    }

    private fun dedupeByCandidate(items: List<Ranked>): List<Ranked> {
        val seen = HashSet<String>()
        val out = ArrayList<Ranked>(items.size)
        for (r in items) {
            val key = TrackDedupe.songKey(r.candidate.track)
            if (seen.add(key)) out.add(r)
        }
        return out
    }

    private data class Weights(val tag: Double, val semantic: Double, val taste: Double, val behavior: Double)

    private fun rankWeights(intent: PetIntent): Weights {
        val hardCount = intent.hardLanguages.size + intent.hardRegions.size +
            intent.hardGenres.size + intent.hardSubGenres.size + intent.hardVocalTypes.size +
            intent.excludeLanguages.size + intent.excludeGenres.size + intent.excludeTags.size
        val vibeCount = intent.softMoods.size + intent.softScenes.size + intent.softTextures.size +
            (if (intent.softEnergy == "any") 0 else 1)
        return when {
            hardCount > 0 -> Weights(0.34, 0.24, 0.12, 0.06)
            vibeCount > 0 -> Weights(0.18, 0.34, 0.16, 0.08)
            else -> Weights(0.08, 0.16, W_TASTE, W_BEHAVIOR)
        }
    }

    private fun hitsAvoid(c: CandidateRecall.Candidate, avoid: List<String>): Boolean {
        val hay = (c.track.title + " " + c.track.artist + " " + c.track.album).lowercase()
        val semanticHay = c.semanticProfile?.let { sp ->
            (sp.genres + sp.subGenres + sp.styleAnchors + sp.moods + sp.scenes +
                sp.textures + sp.energyWords + sp.tempoFeel + sp.negativeTags +
                listOf(sp.language.key, sp.region.key, sp.vocalType.key, sp.summary))
                .joinToString(" ")
                .lowercase()
        }.orEmpty()
        return avoid.any { it in hay || it in semanticHay }
    }

    private fun recentPenalty(
        trackId: Long?,
        in24h: Set<Long>?,
        in7d: Set<Long>?,
        last24hPenalty: Double,
        last7dPenalty: Double,
    ): Double {
        if (trackId == null) return 0.0
        if (in24h?.contains(trackId) == true) return last24hPenalty
        if (in7d?.contains(trackId) == true) return last7dPenalty
        return 0.0
    }

    private fun clamp01(x: Double): Double = if (x < 0) 0.0 else if (x > 1) 1.0 else x
}
